### ISSUE TITLE
0xyPhilic - Withdrawals could be halted temporarily

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2023-04-unitasprotocol-judging#13]
0xyPhilic

medium

# Withdrawals could be halted temporarily

## Summary

The protocol accepts USDT as deposits for which it issues USD1 stable coin based on the current price of USD1/USDT. Later users can convert the USD1 to other stable coins like USD91 (INR) based on the INR/USD price feed. The minimum threshold of funds that have to be present when minting USD1 with USDT is expected to be => 130%, which includes the deposit of the user (i.e. reserves) and the funds stored in `InsurancePool.sol` contract, which essentially are used for overcollateralization. The funds in the `InsurancePool.sol` contract (namely USDT) can be utilized by the protocol to "b/ generate yield via external yield venues" as mentioned in the White paper. This is done by invoking the `sendPortfolio` function which transfers the desired amount from  `InsurancePool.sol`  to a receiver (the guardian as example). The debt that the guardian incurs is tracked via the following mapping 
```solidity
mapping(address => uint256) internal _portfolio;
```
Those funds can be repaid at anytime by invoking `receivePortfolio()` function.

## Vulnerability Detail

The problem lies in the withdrawal flow or when a user wants to swap his USD1 balance and receive USDT. When doing that operation there is the following check:

```solidity
_require(collateral - _getPortfolio(token) >= amount, Errors.POOL_BALANCE_INSUFFICIENT);
```

which is performed in the `withdrawCollateral()` function on the `InsurancePool.sol`. The `withdrawCollateral()` function is utilized in the `Unitas.sol` swap flow whenever the collateral stored in the Unitas.sol contract is insufficient to cover the user expected output of USDT. The idea is that the additional collateral needed is pulled from the `InsurancePool.sol`.

However in a situation where the protocol utilized a big percentage of the funds stored in the `InsurancePool.sol` by using `sendPortfolio()` functionality, there could be insufficient collateral for the user to redeem his USD1 for USDT and the `InsurancePool.sol` might not be able to cover the difference which would block the user from withdrawing his USDT.

Imagine the following situation:

1/ Alice mints 100 USD1 by depositing 100 USDT (considering a price of 1:1)
2/ Alice mints USD91 with her 100 USD1, receiving X USD91 (depending on the price)
3/ There are Y amount of USDT on the `InsurancePool.sol`, which the protocol decides to utilize, thus taking those funds out of the contract
4/ The USD91 (INR) price appreciates in value against the USD significantly (all the expected supported stable coins are quite volatile i.e. INR, TRY etc.)
5/ Alice decides to grab her winning so she converts her X USD91 for Z USD1, which shouldn't be a problem as it is burn/mint operation
6/ Alice then decides to convert her gains to USDT, but since she now has way more USD1 than initially (Z > X), Alice should receive more USDT, but the `reserves` does not have a sufficient amount so the `InsurancePool` will have to cover the difference. However the `InsurancePool` does not currently have the funds as they are utilized for yield generation externally, so Alice is unable to withdraw USDT

Of course the protocol could repay the portfolio and Alice would be able to withdraw the USDT she wants, but her withdrawal is halted temporarily which creates a bad UX.

## Impact

A potentially profitable position (as the Alice example above) could be blocked for withdrawal temporarily due to lack of funds in the `InsurancePool.sol`.

## Code Snippet

Added to avoid warning: https://github.com/sherlock-audit/2023-04-unitasprotocol/blob/main/Unitas-Protocol/src/InsurancePool.sol#L98-L109

## Tool used

Manual Review

## Recommendation

I would suggest setting a `limit` in terms of % on how much can be withdrawn from the `InsurancePool.sol` and utilized for yield generation outside of the protocol. This would make sure that the above mentioned situation is very unlikely to happen, while at the same time protecting the protocol and its' users from a potential exploit on external platform by decreasing the possible impact.

Duplicate of #15

**Comment by 0xyPhilic** [comment]
Escalate for 10 USDC

**Comment by jacksanford1** [comment]
This is an interesting scenario, but also a pretty standard undercollateralization. We'd have to hear how @Adityaxrex intends for the protocol to work, but it seems like the intention would be:

1) There is never a guarantee that there will be enough USDT around to pay for 100% of a profitable trade
2) There is never a guarantee that there will be enough USDT outside of the insurance pool to pay for 100% of a user's expected withdrawal value. 

It seems like these 2 statements would be the intention because there is no way to prevent either from being false (as long as USD91 can differ in price from USD1 and USDT for #1 and the insurance pool can lose money for #2).

If these 2 statements are the intention, then this issue is invalid because it's considered a known risk of the protocol. 

@0xyPhilic can disprove this by showing that either #1 or #2 are not intended to be true, or presenting a different scenario that I'm missing.

**Comment by jacksanford1** [comment]
Ok thanks @0xyPhilic, I think I understand better now. 

> could lead to a withdrawal being halted until some or all the taken funds from the InsurancePool are retrieved, which could take time and might be blocked by external factors not in the scope here.

This impact does not constitute a material loss of funds so the issue cannot be classified as a Medium. The impact is a temporary freezing that can be solved partially by the admin calling `receivePortfolio` and then having the user try again. And any undercollateralization due to price movements or loss of funds in yield strategies likely constitutes a "known risk" so that can't be an issue either (unless you show somewhere that those risks are not acceptable).

**Comment by 0xyPhilic** [comment]
@jacksanford1 I agree that as long as the funds from the `InsurancePool` are not lost due to hack or any other external force outside of the protocol during yield generation, there is no direct loss of funds. It just creates a bad UX for the user as he/she won't be able to withdraw the collateral when he/she needs it. 

A more severe risk here would be in the case of a hack/loss of funds during yield generation in terms that the `InsurancePool` is touched only if there are not enough funds for the withdrawal in the `Unitas` contract, which means that other users would essentially collect their winnings from the funds used by other users that minted USD1. In the case of a bank run the last user would have a close to or even 100% loss of funds as there won't be any USDT left in the `Unitas` contract and the `InsurancePool` would also be depleted. I don't know if this can be considered an issue or part of the problem expressed here regarding the `receivePortfolio` functionality.

**Comment by jacksanford1** [comment]
@0xyPhilic This is a very good point. There are many protocols on-chain that actually operate using this logic (all users get 100% of funds back even if the protocol is undercollateralized, then the last users get 0% back). 

I would assume Unitas also plans to work this way because it's common (and the code points in this direction). But if this is not the intended mechanism design, then maybe @Adityaxrex @SealCaoXREX can let us know.

**Comment by jacksanford1** [comment]
Update: Unitas does not plan to work this way according to this sentence:

> The protocol allows unrestricted and unconditional exit from EMCs back to USDPEGA – that is,
conversions from USD_EMC to USD1 and then from USD1 to USDPEGA.

If there are no USDT left for the last user, then the unconditional exit assumption is broken. 

This will be duped in the "breaks unconditional exit" family.

**Comment by jacksanford1** [comment]
Escalation accepted

Duplicate of #95

**Comment by jacksanford1** [comment]
Result:
Medium
Has Duplicates

**Comment by hrishibhat** [comment]
Result:
Medium
Duplicate of #95

**Comment by sherlock-admin** [comment]
Escalations have been resolved successfully!

Escalation status:
- [0xyPhilic](https://github.com/sherlock-audit/2023-04-unitasprotocol-judging/issues/13/#issuecomment-1603910064): accepted
